File: test/system/console_test.rb
require "application_system_test_case"

class ConsoleSystemTest < ApplicationSystemTestCase
  setup do
    # Ensure we're in development environment for tests
    Rails.env = "development"
  end

  test "visiting the console page" do
    visit console_path
    
    assert_selector "h1", text: "Rails Web Console", visible: false
    assert_selector ".console-container"
    assert_selector "#console-input"
    assert_selector "#console-output"
    assert_selector ".console-header"
  end

  test "console input is focused on load" do
    visit console_path
    
    # Check that the input field has focus
    input_element = find("#console-input")
    assert_equal input_element, page.driver.browser.switch_to.active_element
  end

  test "executing simple commands" do
    visit console_path
    
    # Enter a simple command
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    
    # Wait for the result to appear
    assert_text "2", wait: 5
    assert_text "> 1 + 1"
  end

  test "executing Rails model commands" do
    # Create test data
    Post.create!(title: "Test Post", body: "Test body")
    
    visit console_path
    
    # Test Post.count
    fill_in "console-input", with: "Post.count"
    find("#console-input").send_keys(:return)
    
    # Should show count result
    assert_text "1", wait: 5
    assert_text "> Post.count"
  end

  test "command history navigation" do
    visit console_path
    
    # Execute first command
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    
    # Wait for completion
    assert_text "2", wait: 5
    
    # Execute second command
    fill_in "console-input", with: "2 + 2"
    find("#console-input").send_keys(:return)
    
    # Wait for completion
    assert_text "4", wait: 5
    
    # Navigate history with arrow keys
    find("#console-input").send_keys(:arrow_up)
    assert_equal "2 + 2", find("#console-input").value
    
    find("#console-input").send_keys(:arrow_up)
    assert_equal "1 + 1", find("#console-input").value
    
    find("#console-input").send_keys(:arrow_down)
    assert_equal "2 + 2", find("#console-input").value
  end

  test "error handling display" do
    visit console_path
    
    # Execute a command that will cause an error
    fill_in "console-input", with: "undefined_variable"
    find("#console-input").send_keys(:return)
    
    # Should display error
    assert_text "Error:", wait: 5
    assert_text "NameError"
  end

  test "clear output button works" do
    visit console_path
    
    # Execute some commands to populate output
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    assert_text "2", wait: 5
    
    fill_in "console-input", with: "2 + 2"
    find("#console-input").send_keys(:return)
    assert_text "4", wait: 5
    
    # Click clear output button
    click_button "Clear Output"
    
    # Commands should be cleared but welcome messages should remain
    assert_text "Welcome to Rails Web Console"
    assert_no_text "> 1 + 1"
    assert_no_text "> 2 + 2"
  end

  test "clear history button works" do
    visit console_path
    
    # Execute some commands
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    assert_text "2", wait: 5
    
    # Accept the confirmation dialog and clear history
    accept_confirm do
      click_button "Clear History"
    end
    
    # Should show history cleared message
    assert_text "Command history cleared", wait: 5
  end

  test "keyboard shortcuts work" do
    visit console_path
    
    # Execute a command to populate output
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    assert_text "2", wait: 5
    
    # Test Ctrl+L to clear output
    find("#console-input").send_keys([:control, 'l'])
    
    # Output should be cleared
    assert_text "Welcome to Rails Web Console"
    assert_no_text "> 1 + 1"
  end

  test "execute button works" do
    visit console_path
    
    # Enter command and click execute button instead of pressing enter
    fill_in "console-input", with: "3 + 3"
    click_button "Execute"
    
    # Should show result
    assert_text "6", wait: 5
    assert_text "> 3 + 3"
  end

  test "helper commands work" do
    visit console_path
    
    # Test models helper
    fill_in "console-input", with: "models"
    find("#console-input").send_keys(:return)
    
    # Should show available models
    assert_text "Post", wait: 5
    assert_text "> models"
  end

  test "clicking on history items reuses commands" do
    visit console_path
    
    # Execute a command first
    fill_in "console-input", with: "5 + 5"
    find("#console-input").send_keys(:return)
    assert_text "10", wait: 5
    
    # If there are history items displayed, click on one
    if page.has_css?(".history-item")
      first(".history-item").click
      # The input should be filled with the clicked command
      assert_not_equal "", find("#console-input").value
    end
  end

  test "console shows welcome messages and examples" do
    visit console_path
    
    # Check for welcome messages
    assert_text "Welcome to Rails Web Console"
    assert_text "Type Ruby/Rails commands below"
    assert_text "Try these commands:"
    
    # Check for example commands
    assert_text "models"
    assert_text "routes"
    assert_text "Post.count"
    assert_text "model_info(Post)"
    assert_text "env_info"
  end

  test "console handles long output gracefully" do
    visit console_path
    
    # Execute a command that produces long output
    fill_in "console-input", with: "(1..100).to_a"
    find("#console-input").send_keys(:return)
    
    # Should handle the long array output
    assert_text "more items", wait: 10
  end

  test "console input clears after execution" do
    visit console_path
    
    # Enter and execute command
    fill_in "console-input", with: "1 + 1"
    find("#console-input").send_keys(:return)
    
    # Wait for execution
    assert_text "2", wait: 5
    
    # Input should be cleared
    assert_equal "", find("#console-input").value
  end

  test "loading indicator appears during execution" do
    visit console_path
    
    # Execute command and check for loading indicator
    fill_in "console-input", with: "sleep(0.1); 'done'"
    find("#console-input").send_keys(:return)
    
    # Should briefly show loading message
    # Note: This might be too fast to reliably test in some environments
    # assert_text "Executing...", wait: 1
    
    # Should eventually show result
    assert_text "done", wait: 5
  end

  private

  def console_path
    "/console"
  end
end

File: test/javascript/console_test.js
// JavaScript Test Scenarios for Rails Web Console
// These are manual test scenarios since Rails doesn't have built-in JS unit testing

/*
Test Scenarios for JavaScript Functionality:

1. Console Initialization:
   - Page loads with console interface
   - Input field is automatically focused
   - Welcome messages are displayed
   - Command history is loaded from session

2. Command Execution:
   - Typing command and pressing Enter executes it
   - Click on Execute button works
   - Command is added to output with prompt
   - Loading indicator appears during execution
   - Result is displayed after completion
   - Input field is cleared after execution

3. Command History:
   - Up arrow navigates to previous commands
   - Down arrow navigates to next commands
   - History wraps around correctly
   - Command history persists across sessions

4. Error Handling:
   - Network errors are displayed properly
   - Server errors are formatted correctly
   - Loading indicators are removed on error

5. UI Interactions:
   - Clear Output button removes command output
   - Clear History button clears command history (with confirmation)
   - Clicking on history items fills input field
   - Keyboard shortcuts work (Ctrl+L)

6. Output Formatting:
   - Arrays are displayed as separate lines
   - Long strings are truncated appropriately
   - Objects are formatted readably
   - Timestamps are shown for each command

To run these tests manually:

1. Open browser console (F12)
2. Navigate to /console
3. Execute each test scenario
4. Verify expected behavior

Example test commands:
- Simple: 1 + 1
- String: "hello world"
- Array: [1, 2, 3, 4, 5]
- Error: undefined_variable
- Rails: Post.count
- Helper: models

Automated testing could be added using:
- Jasmine
- QUnit  
- Jest (with Rails integration)
- Capybara with JavaScript driver
*/

// Test helper functions that could be added to the console.js file
function runJavaScriptTests() {
  console.log("Starting JavaScript Console Tests...");
  
  // Test 1: Console object exists
  if (typeof window.railsConsole !== 'undefined') {
    console.log("✓ Console object initialized");
  } else {
    console.error("✗ Console object not found");
  }
  
  // Test 2: Required DOM elements exist
  const requiredElements = ['console-input', 'console-output'];
  requiredElements.forEach(id => {
    if (document.getElementById(id)) {
      console.log(`✓ Element ${id} found`);
    } else {
      console.error(`✗ Element ${id} missing`);
    }
  });
  
  // Test 3: Event listeners are attached
  const inputEl = document.getElementById('console-input');
  if (inputEl && inputEl.onkeydown) {
    console.log("✓ Input event listeners attached");
  } else {
    console.log("? Input event listeners may not be attached");
  }
  
  // Test 4: Command history array exists
  if (Array.isArray(window.initialHistory)) {
    console.log("✓ Command history initialized");
  } else {
    console.error("✗ Command history not initialized");
  }
  
  console.log("JavaScript Console Tests completed. Check console for results.");
}

// Performance test helper
function measureConsolePerformance() {
  const startTime = performance.now();
  
  // Simulate command execution
  if (typeof executeCommand === 'function') {
    // This would need to be adapted to not actually execute
    console.log("Console performance test placeholder");
  }
  
  const endTime = performance.now();
  console.log(`Console operation took ${endTime - startTime} milliseconds`);
}

// Memory leak test helper
function checkMemoryLeaks() {
  const initialMemory = performance.memory ? performance.memory.usedJSHeapSize : 0;
  
  // Simulate multiple command executions
  for (let i = 0; i < 100; i++) {
    // Add fake output lines
    const outputEl = document.getElementById('console-output');
    if (outputEl) {
      const line = document.createElement('div');
      line.textContent = `Test line ${i}`;
      outputEl.appendChild(line);
    }
  }
  
  // Clear the test lines
  const testLines = document.querySelectorAll('#console-output div');
  testLines.forEach(line => {
    if (line.textContent.startsWith('Test line')) {
      line.remove();
    }
  });
  
  const finalMemory = performance.memory ? performance.memory.usedJSHeapSize : 0;
  const memoryDiff = finalMemory - initialMemory;
  
  console.log(`Memory usage difference: ${memoryDiff} bytes`);
  
  if (memoryDiff < 100000) { // Less than 100KB difference
    console.log("✓ No significant memory leaks detected");
  } else {
    console.warn("? Potential memory leak detected");
  }
}

// Add these functions to window for manual testing
if (typeof window !== 'undefined') {
  window.runJavaScriptTests = runJavaScriptTests;
  window.measureConsolePerformance = measureConsolePerformance;
  window.checkMemoryLeaks = checkMemoryLeaks;
}

/*
Manual Testing Checklist:

□ Open /console in browser
□ Check browser console for JavaScript errors
□ Run: runJavaScriptTests() in browser console
□ Test command execution: type "1 + 1" and press Enter
□ Test history: execute multiple commands, use up/down arrows
□ Test error handling: type "undefined_variable"
□ Test clear buttons: click "Clear Output" and "Clear History"
□ Test keyboard shortcuts: Ctrl+L to clear
□ Test on different browsers: Chrome, Firefox, Safari, Edge
□ Test on mobile devices if applicable
□ Run: measureConsolePerformance() for performance check
□ Run: checkMemoryLeaks() for memory leak detection

Performance Benchmarks:
- Command execution: < 100ms response time
- Page load: < 2 seconds
- Memory usage: < 50MB for typical session
- History navigation: < 10ms

Compatibility Requirements:
- Modern browsers (ES6+ support)
- JavaScript enabled
- Fetch API support
- CSS Grid/Flexbox support
*/
